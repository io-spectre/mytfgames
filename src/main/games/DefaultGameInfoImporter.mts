import { inject, injectable } from "inversify";
import { Temporal } from "temporal-polyfill";
import * as uuid from "uuid";

import { BundledGameMetadata, GameInfoLoader } from "$game-info/loader";
import { Game, gameCrawled, GameSId, ImportService } from "$ipc/main-renderer";
import { DatabaseProvider } from "$node-base/database";
import { RemoteReduxActionSender } from "$node-base/ipc";
import { remoteProcedure } from "$pure-base/ipc";

@injectable()
export class DefaultGameInfoImporter {
  constructor(
    @inject(DatabaseProvider) private readonly db: DatabaseProvider,
    @inject(RemoteReduxActionSender)
    private readonly remoteRedux: RemoteReduxActionSender,
    @inject(GameInfoLoader) private readonly gameInfoLoader: GameInfoLoader,
  ) {}

  @remoteProcedure(ImportService, "importGameInfoFromFolder")
  async importGameInfoFromFolder(artifactFolder: string): Promise<GameSId> {
    let metadata: BundledGameMetadata;
    try {
      metadata =
        await this.gameInfoLoader.loadGameInfoFromMetaFolder(artifactFolder);
    } catch (error) {
      throw new Error(`Failed to load game info from folder`, { cause: error });
    }
    const game = await this.db.transaction().execute(async (trx) => {
      await trx
        .insertInto("game")
        .values({
          game_id: metadata.gameInfo.ids.uuid,
        })
        .onConflict((oc) => oc.doNothing())
        .execute();

      const updatedMetadata = await trx
        .insertInto("game_metadata")
        .values({
          game_id: metadata.gameInfo.ids.uuid,
          name: metadata.gameInfo.name,
          synopsis: metadata.gameInfo.synopsis,
          full_description: metadata.gameInfo.description,
          last_update_datetime: Temporal.Now.instant().toString({
            smallestUnit: "second",
          }),
          metadata_timestamp: metadata.gameInfo.metadataTimestamp,
          tfgames_site_game_id: metadata.gameInfo.ids.tfgamesSiteGameId,
        })
        .onConflict((oc) =>
          oc
            .doUpdateSet((eb) => ({
              game_id: eb.ref("excluded.game_id"),
              name: eb.ref("excluded.name"),
              synopsis: eb.ref("excluded.synopsis"),
              full_description: eb.ref("excluded.full_description"),
              last_update_datetime: eb.ref("excluded.last_update_datetime"),
              metadata_timestamp: eb.ref("excluded.metadata_timestamp"),
              tfgames_site_game_id: eb.ref("excluded.tfgames_site_game_id"),
            }))
            .whereRef(
              "excluded.metadata_timestamp",
              "<",
              "game_metadata.metadata_timestamp",
            ),
        )
        .returning([
          "name as name",
          "last_update_datetime as lastUpdateTimestamp",
          "synopsis as synopsis",
          "full_description as fullDescription",
        ])
        .executeTakeFirst();

      await Promise.all(
        metadata.versionInfo.map((version) => {
          return trx
            .insertInto("game_version")
            .values({
              game_id: metadata.gameInfo.ids.uuid,
              version: version.version,
              note: version.note,
            })
            .onConflict((oc) =>
              oc
                .doUpdateSet({ note: version.note })
                .where("game_version.note", "=", ""),
            )
            .execute();
        }),
      );

      const userNotes = await trx
        .selectFrom("game_user_notes")
        .where("game_id", "=", metadata.gameInfo.ids.uuid)
        .select([
          "custom_name",
          "last_played_datetime",
          "last_change_datetime",
          "user_rating",
          "note",
        ])
        .executeTakeFirst();
      const listing = await trx
        .selectFrom("game_official_listing")
        .where("game_id", "=", metadata.gameInfo.ids.uuid)
        .select([
          "tfgames_game_id as tfgamesId",
          "name",
          "num_likes as numLikes",
          "last_update_datetime as lastUpdateTimestamp",
        ])
        .executeTakeFirst();

      return {
        id: uuid.stringify(metadata.gameInfo.ids.uuid),
        userNotes: userNotes
          ? {
              name: userNotes.custom_name,
              lastChangeTimestamp: userNotes.last_change_datetime,
              lastPlayedTimestamp: userNotes.last_played_datetime,
              userRating: userNotes.user_rating,
              note: userNotes.note,
            }
          : null,
        metadata: updatedMetadata ?? null,
        listing: listing ?? null,
      } satisfies Game;
    });
    this.remoteRedux.dispatch(gameCrawled(game));
    return game.id;
  }
}
