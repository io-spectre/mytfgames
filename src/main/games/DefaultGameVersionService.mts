import { inject, injectable } from "inversify";
import * as R from "remeda";
import * as uuid from "uuid";

import {
  GameSId,
  GameVersion,
  GameVersionService as GameVersionServiceContract,
} from "$ipc/main-renderer";
import { DatabaseProvider, GameId } from "$node-base/database";
import { remoteProcedure } from "$pure-base/ipc";

@injectable()
export class DefaultGameVersionService {
  constructor(
    @inject(DatabaseProvider) private readonly db: DatabaseProvider,
  ) {}

  @remoteProcedure(GameVersionServiceContract, "retrieveVersionsForGame")
  async retrieveVersionsForGame(gameId: GameSId): Promise<GameVersion[]> {
    const id = uuid.parse(gameId) as GameId;

    const [versions, sources, artifacts] = await this.db
      .transaction()
      .execute(async (trx) =>
        Promise.all([
          trx
            .selectFrom("game_version")
            .select(["version", "note"])
            .where("game_id", "=", id)
            .orderBy("version", "desc")
            .execute(),
          trx
            .selectFrom("game_version_source")
            .select(["version", "uri", "official_note as officialNote"])
            .where("game_id", "=", id)
            .execute(),
          trx
            .selectFrom("game_version_artifact")
            .select([
              "version",
              "platform_type as platform",
              "node_no as rootNodeNo",
            ])
            .where("game_id", "=", id)
            .execute(),
        ]),
      );

    const groupedSources = R.groupBy(sources, (s) => s.version);
    const groupedArtifacts = R.groupBy(artifacts, (a) => a.version);
    return versions.map(({ note, version }) => ({
      gameId,
      version,
      note,
      sources: groupedSources[version] ?? [],
      artifacts: groupedArtifacts[version] ?? [],
    }));
  }
}
