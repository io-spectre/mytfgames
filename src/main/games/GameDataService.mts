import { GameId, TfgamesGameId } from "$node-libs/database";
import { makeServiceIdentifier } from "$node-libs/utils";

const GameDataService =
  makeServiceIdentifier<GameDataService>("game info service");
interface GameDataService {
  downloadGameInfo(id: number): Promise<void>;
  refreshIndex(): Promise<[GameId, TfgamesGameId][]>;
}
export { GameDataService };
