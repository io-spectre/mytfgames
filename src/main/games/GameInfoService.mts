import { GameInfo, GameList, GameOrderType } from "$ipc/main-renderer";
import { makeServiceIdentifier } from "$main/utils";

const GameInfoService =
  makeServiceIdentifier<GameInfoService>("game info service");
interface GameInfoService {
  getGames(ids: string[]): Promise<GameInfo[]>;
  getGameList(
    order: GameOrderType,
    page: number,
    pageSize: number,
    force: boolean,
  ): Promise<GameList>;
}
export { GameInfoService };
