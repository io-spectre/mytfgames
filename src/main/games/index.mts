export * from "./GameDataService.mjs";
export * from "./GameVersionService.mjs";
