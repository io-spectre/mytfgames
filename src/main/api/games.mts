import { Cheerio, Element, load as cheerioLoad } from "cheerio";
import { net } from "electron/main";
import { injectable } from "inversify";

import { GameInfo } from "$ipc/main-renderer";
import { makeServiceIdentifier } from "$main/utils";

const idRegex = /&id=(\d+)/;

type PropExtractor = ($td: Cheerio<Element>) => Partial<GameInfo> | undefined;
const propExtractors = Object.assign(
  Object.create(null) as Record<string, PropExtractor | undefined>,
  {
    Name: ($td: Cheerio<Element>) => ({
      rid: idRegex.exec($td.find("a").get(0)!.attribs["href"] ?? "")![1]!,
      name: $td.text().trim(),
    }),
    "Last Update": ($td: Cheerio<Element>) => ({
      lastUpdate: $td.text().trim(),
    }),
  },
);

const GamesApi = makeServiceIdentifier<GamesApi>("games api");
interface GamesApi {
  getGames(): Promise<GameInfo[]>;
}
export { GamesApi };

@injectable()
export class GamesApiImpl {
  /**
   * Note that this call is not cached and _very_ expensive on the server side.
   */
  async getGames(): Promise<GameInfo[]> {
    const endpoint = "https://tfgames.site/index.php";
    const requestParams = new URLSearchParams();
    (
      [
        ["module", "search"],
        ["search", "1"],
        ["searchcontent", ""],
        // We need to specify at least one search parameter in order to get
        // any results at all. So I have excluded games via "author" using an
        // account that has not been online for a long time and whose games are
        // no longer available on the website.
        // I'm sorry "Dude" but you're the chosen one.
        // "3224" => "Dude"; 1 very old RAGS game; not online for over a decade
        ["exauthor[]", "3224"],
        ["likesmin", "0"],
        ["likesmax", "0"],
      ] as const
    ).forEach(([key, value]) => requestParams.set(key, value));

    const response = await net.fetch(endpoint, {
      method: "POST",
      headers: {},
      body: requestParams,
    });

    return this.parseGameList(await response.text());
  }

  parseGameList(response: string): GameInfo[] {
    const $ = cheerioLoad(response);
    const $searchresults = $("div.searchresultcontainer > table");

    const $headers = $searchresults.find("thead > tr > th");
    // double map as cheerio removes undefined / null values
    const columnPropExtractor = $headers
      .map((_, th) => $(th).text().trim())
      .get()
      .map((headerText) => propExtractors[headerText]);

    const games = $searchresults
      .find("tbody > tr")
      .map((_, tr) => {
        const $tds = $(tr).find("td");
        const partials = $tds
          .map((i, td) => columnPropExtractor[i]?.($(td)))
          .get();
        // TODO: validate that all props are defined
        return Object.assign({}, ...partials) as GameInfo;
      })
      .get();

    return games;
  }
}
