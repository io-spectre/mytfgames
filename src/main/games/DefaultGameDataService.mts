import { inject, injectable } from "inversify";
import { Insertable, QueryExecutorProvider, sql } from "kysely";
import * as R from "remeda";
import { Temporal } from "temporal-polyfill";
import * as uuid from "uuid";

import {
  Game,
  gameCrawled,
  GameDataService as GameDataServiceContract,
  GameOrderType,
  GameSId,
} from "$ipc/main-renderer";
import { GamesApi, RemoteCategory, RemoteVersionInfo } from "$main/api";
import { RemoteReduxActionSender } from "$main/pal";
import {
  AppQueryCreator,
  DatabaseProvider,
  GameId,
  GameTagTable,
  TagTable,
  TfgamesGameId,
  WellKnownTagCategory,
} from "$node-base/database";
import { remoteProcedure } from "$pure-base/ipc";

import { GameDataService } from "./GameDataService.mjs";

@injectable()
export class DefaultGameDataService implements GameDataService {
  constructor(
    @inject(DatabaseProvider) private readonly db: DatabaseProvider,
    @inject(GamesApi) private readonly gamesApi: GamesApi,
    @inject(RemoteReduxActionSender)
    private readonly remoteRedux: RemoteReduxActionSender,
  ) {}

  @remoteProcedure(GameDataServiceContract, "retrieveOrder")
  async retrieveOrder(): Promise<Record<GameOrderType, GameSId[]>> {
    return await this.db.transaction().execute(async (trx) => {
      const partials = await Promise.all(
        Object.values(GameOrderType).map(
          async (type) =>
            [
              type,
              (await this.retrieveOrderForType(trx, type)).map((r) =>
                uuid.stringify(r.id),
              ),
            ] satisfies [GameOrderType, GameSId[]],
        ),
      );
      return Object.fromEntries(partials) as Record<GameOrderType, GameSId[]>;
    });
  }

  private async retrieveOrderForType(
    trx: AppQueryCreator,
    type: GameOrderType,
  ): Promise<{ id: Buffer }[]> {
    const ids = trx
      .selectFrom("game")
      .select(["game.game_id as id"])
      .leftJoin(
        "game_description as description",
        "description.game_id",
        "game.game_id",
      )
      .leftJoin(
        "game_official_listing as listing",
        "listing.game_id",
        "game.game_id",
      )
      .leftJoin(
        "game_official_blacklist as blacklist",
        "blacklist.game_id",
        "game.game_id",
      )
      .where("blacklist.last_crawl_datetime", "is", null);
    switch (type) {
      case GameOrderType.Id:
        return await ids.orderBy("game.game_id").execute();
      case GameOrderType.Name:
        return await ids
          .orderBy((eb) => eb.fn.coalesce("description.name", "listing.name"))
          .execute();
      case GameOrderType.LastUpdate:
        return await ids
          .orderBy((eb) =>
            eb.fn.coalesce(
              "listing.last_update_datetime",
              "description.last_change_datetime",
            ),
          )
          .execute();
    }
  }

  @remoteProcedure(GameDataServiceContract, "retrieveGamesById")
  async getGames(ids: string[]): Promise<Game[]> {
    const chunks = R.chunk(ids.map(uuid.parse) as GameId[], 512);

    const chunkResults = await this.db
      .transaction()
      .execute(
        async (trx) =>
          await Promise.all(
            chunks.map((c) =>
              trx
                .selectFrom("game")
                .leftJoin(
                  "game_official_listing as listing",
                  "listing.game_id",
                  "game.game_id",
                )
                .leftJoin(
                  "game_description as description",
                  "description.game_id",
                  "game.game_id",
                )
                .select([
                  "game.game_id as id",
                  "description.name as name",
                  "description.last_change_datetime as lastChange",
                  "description.last_played_datetime as lastPlayed",
                  "description.note as note",
                  "description.user_rating as userRating",
                  "listing.name as officialName",
                  "listing.num_likes as numLikes",
                  "listing.tfgames_game_id as tfgamesId",
                  "listing.last_update_datetime as lastUpdate",
                ])
                .where("game.game_id", "in", c)
                .execute(),
            ),
          ),
      );

    return R.pipe(
      chunkResults,
      R.flat(),
      R.map(
        ({
          id,
          lastChange,
          lastPlayed,
          lastUpdate,
          name,
          note,
          numLikes,
          officialName,
          tfgamesId,
          userRating,
        }) =>
          ({
            id: uuid.stringify(id),
            description:
              name == null
                ? null
                : {
                    name,
                    lastChangeTimestamp: lastChange!,
                    lastPlayedTimestamp: lastPlayed!,
                    userRating: userRating!,
                    note: note!,
                  },
            listing:
              tfgamesId == null
                ? null
                : {
                    tfgamesId,
                    name: officialName!,
                    numLikes: numLikes!,
                    lastUpdateTimestamp: lastUpdate!,
                  },
          }) satisfies Game,
      ),
    );
  }

  @remoteProcedure(GameDataServiceContract, "findGamesByNamePrefix")
  async findGameByNamePrefix(prefix: string): Promise<GameSId[]> {
    const games = await this.db
      .selectFrom("game")
      .leftJoin("game_description", "game_description.game_id", "game.game_id")
      .leftJoin(
        "game_official_listing",
        "game_official_listing.game_id",
        "game.game_id",
      )
      .where((eb) =>
        eb.or([
          eb("game_official_listing.name", "like", `${prefix}%`),
          eb("game_description.name", "like", `${prefix}%`),
        ]),
      )
      .select(["game.game_id"])
      .execute();
    return games.map((r) => uuid.stringify(r.game_id));
  }

  @remoteProcedure(GameDataServiceContract, "updateGameDescription")
  async updateGameDescription(
    id: GameSId,
    description: {
      name: string;
      userRating: number;
      note: string;
    },
  ): Promise<void> {
    const game_id = uuid.parse(id) as GameId;
    const last_change_datetime = Temporal.Now.instant().toString({
      smallestUnit: "second",
    });
    const game = await this.db.transaction().execute(async (trx) => {
      const updatedDescription = await trx
        .insertInto("game_description")
        .values({
          game_id,
          name: description.name,
          user_rating: description.userRating,
          note: description.note,
          last_change_datetime,
        })
        .onConflict((oc) =>
          oc.doUpdateSet({
            last_change_datetime,
            name: description.name,
            user_rating: description.userRating,
            note: description.note,
          }),
        )
        .returning([
          "name",
          "last_change_datetime as lastChangeTimestamp",
          "last_played_datetime as lastPlayedTimestamp",
          "user_rating as userRating",
          "note",
        ])
        .executeTakeFirstOrThrow();

      const listing = await trx
        .selectFrom("game_official_listing")
        .where("game_id", "=", game_id)
        .select([
          "tfgames_game_id as tfgamesId",
          "name",
          "num_likes as numLikes",
          "last_update_datetime as lastUpdateTimestamp",
        ])
        .executeTakeFirst();

      return {
        id,
        description: updatedDescription,
        listing: listing ?? null,
      };
    });
    this.remoteRedux.dispatch(gameCrawled(game));
  }

  async refreshIndex(): Promise<[GameId, TfgamesGameId][]> {
    const currentIndex = await this.gamesApi.getGames();

    const updated = await this.db.transaction().execute(async (trx) => {
      await Promise.all(
        currentIndex.map(async ({ id, name, numLikes, lastUpdate }) => {
          const game = await trx
            .updateTable("game_official_listing")
            .where("tfgames_game_id", "=", id)
            .set("name", name)
            .set("last_update_datetime", lastUpdate)
            .returning("game_id")
            .executeTakeFirst();
          if (game === undefined) {
            await this.insertOfficialListing(
              trx,
              null,
              id,
              name,
              numLikes,
              lastUpdate,
            );
          }
        }),
      );
      return await trx
        .selectFrom("game_official_listing_outdated_v")
        .select(["game_id", "tfgames_game_id"])
        .execute();
    });
    return updated.map(({ game_id, tfgames_game_id }) => [
      game_id,
      tfgames_game_id,
    ]);
  }

  private async insertOfficialListing(
    qc: AppQueryCreator,
    gameId: GameId | null | undefined,
    tfgames_game_id: number,
    name: string,
    num_likes: number,
    last_update_datetime: string,
  ): Promise<Buffer> {
    const game_id = gameId ?? uuid.v4(null, Buffer.allocUnsafe(16));
    if (gameId == null) {
      await qc.insertInto("game").values({ game_id }).execute();
    }
    await qc
      .insertInto("game_official_listing")
      .values({
        game_id,
        tfgames_game_id,
        name,
        num_likes,
        last_update_datetime,
      })
      .execute();
    return game_id;
  }

  async downloadGameInfo(tfgamesGameId: TfgamesGameId): Promise<void> {
    const gameDetails = await this.gamesApi.getGameDetails(tfgamesGameId);

    const game = await this.db.transaction().execute(async (trx) => {
      const lastCrawled = Temporal.Now.instant().toString({
        smallestUnit: "second",
      });
      if (!("name" in gameDetails)) {
        const game = await trx
          .selectFrom("game_official_listing")
          .select("game_id")
          .where("tfgames_game_id", "=", tfgamesGameId)
          .executeTakeFirst();
        if (game != null) {
          await trx
            .insertInto("game_official_blacklist")
            .values({
              game_id: game.game_id,
              last_crawl_datetime: lastCrawled,
            })
            .onConflict((oc) =>
              oc.doUpdateSet({ last_crawl_datetime: lastCrawled }),
            )
            .execute();
        }
        return;
      }

      const game_id =
        (
          await trx
            .updateTable("game_official_listing")
            .where("tfgames_game_id", "=", tfgamesGameId)
            .set("name", gameDetails.name)
            .set("last_update_datetime", gameDetails.lastUpdate)
            .set("num_likes", gameDetails.numLikes ?? 0)
            .returning("game_id")
            .executeTakeFirst()
        )?.game_id ??
        (await this.insertOfficialListing(
          trx,
          null,
          tfgamesGameId,
          gameDetails.name,
          gameDetails.numLikes ?? 0,
          gameDetails.lastUpdate,
        ));

      await trx
        .deleteFrom("game_official_blacklist")
        .where("game_id", "=", game_id)
        .execute();

      await trx
        .insertInto("game_official_listing_details")
        .values({
          game_id,
          last_crawl_datetime: lastCrawled,
          release_datetime: gameDetails.releaseDate,
          synopsis: "",
        })
        .onConflict((oc) =>
          oc.column("game_id").doUpdateSet({
            last_crawl_datetime: lastCrawled,
            release_datetime: gameDetails.releaseDate,
          }),
        )
        .execute();

      await Promise.all(
        gameDetails.authors.map((author) =>
          trx
            .insertInto("tfgames_author")
            .values({
              tfgames_profile_id: author.id,
              username: author.name,
            })
            .onConflict((oc) =>
              oc.column("tfgames_profile_id").doUpdateSet({
                username: author.name,
              }),
            )
            .execute(),
        ),
      );
      await Promise.all(
        gameDetails.authors.map((author) =>
          trx
            .insertInto("game_official_tfgames_author")
            .values({
              game_id,
              tfgames_profile_id: author.id,
            })
            .onConflict((oc) =>
              oc.columns(["game_id", "tfgames_profile_id"]).doNothing(),
            )
            .execute(),
        ),
      );

      await this.mergeTags(trx, game_id, [
        [gameDetails.adultThemes, WellKnownTagCategory.Adult],
        [gameDetails.transformationThemes, WellKnownTagCategory.Transformation],
        [gameDetails.multimediaThemes, WellKnownTagCategory.Multimedia],
      ]);

      await this.mergeVersions(trx, game_id, gameDetails.versions);

      return {
        id: uuid.stringify(game_id),
        description: await this.getDescrpition(trx, game_id),
        listing: {
          tfgamesId: tfgamesGameId,
          name: gameDetails.name,
          numLikes: gameDetails.numLikes ?? 0,
          lastUpdateTimestamp: gameDetails.lastUpdate,
        },
      };
    });

    if (game != null) {
      this.remoteRedux.dispatch(gameCrawled(game));
    }
  }

  private async getDescrpition(
    qc: AppQueryCreator & QueryExecutorProvider,
    game_id: GameId,
  ): Promise<Game["description"] | null> {
    return (
      (await qc
        .selectFrom("game_description")
        .where("game_id", "=", game_id)
        .select([
          "name",
          "last_change_datetime as lastChangeTimestamp",
          "last_played_datetime as lastPlayedTimestamp",
          "user_rating as userRating",
          "note",
        ])
        .executeTakeFirst()) ?? null
    );
  }

  private async mergeTags(
    qc: AppQueryCreator & QueryExecutorProvider,
    game_id: GameId,
    categorizedTags: [RemoteCategory[], string][],
  ): Promise<void> {
    if (categorizedTags.length === 0) {
      return;
    }
    const tags = categorizedTags.flatMap(([remoteCategories, tag_category]) =>
      remoteCategories.map(
        ({ abbreviation: tag, name: tag_name }) =>
          ({
            tag_category,
            tag,
            tag_name,
          }) satisfies Insertable<TagTable>,
      ),
    );
    if (tags.length === 0) {
      return;
    }
    await qc
      .insertInto("tag")
      .values(tags)
      .onConflict((oc) => oc.doNothing())
      .execute();

    await Promise.all(
      categorizedTags.flatMap(([remoteCategories, tag_category]) =>
        remoteCategories.map(
          async ({ abbreviation: tag, id: tfgames_tag_id }) =>
            // TODO: replace with insertInto when kysely supports orReplace()
            //       https://github.com/kysely-org/kysely/issues/916
            sql`INSERT OR REPLACE INTO official_tag VALUES (${tag_category}, ${tag}, ${tfgames_tag_id})`.execute(
              qc,
            ),
        ),
      ),
    );

    const gameTags = categorizedTags.flatMap(
      ([remoteCategories, tag_category]) =>
        remoteCategories.map(
          ({ abbreviation: tag }) =>
            ({
              tag_category,
              tag,
              game_id,
            }) satisfies Insertable<GameTagTable>,
        ),
    );
    await qc
      .insertInto("game_tag")
      .values(gameTags)
      .onConflict((oc) => oc.doNothing())
      .execute();
  }

  private async mergeVersions(
    qc: AppQueryCreator,
    game_id: GameId,
    versions: RemoteVersionInfo[],
  ): Promise<void> {
    if (versions.length === 0) {
      return;
    }
    await qc
      .insertInto("game_version")
      .values(versions.map(({ version }) => ({ version, game_id })))
      .onConflict((oc) => oc.doNothing())
      .execute();

    await qc
      .deleteFrom("game_version_source")
      .where("game_id", "=", game_id)
      .where(
        "version",
        "in",
        versions.map(({ version }) => version),
      )
      .execute();

    const values = versions.flatMap(({ version, downloads }) =>
      R.uniqueBy(downloads, (d) => d.link).map(({ link, note }) => ({
        version,
        game_id,
        uri: link,
        official_note: note,
      })),
    );
    if (values.length === 0) {
      return;
    }
    await qc.insertInto("game_version_source").values(values).execute();
  }
}
