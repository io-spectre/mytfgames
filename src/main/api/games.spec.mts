import * as fs from "node:fs";
import path from "node:path";

import { describe, expect, it, vi } from "vitest";

import { GamesApiImpl } from "./games.mjs";

vi.mock("electron/main", () => ({ net: null }));

describe("Games API", () => {
  it("should parse a list response", () => {
    const response = fs.readFileSync(
      path.join(import.meta.dirname, "test/search-multimedia-32.html"),
      "utf8",
    );

    const subject = new GamesApiImpl();
    const result = subject.parseGameList(response);
    expect(result).toMatchSnapshot();
  });
  it("should parse a viewgame response", () => {
    const response = fs.readFileSync(
      path.join(import.meta.dirname, "test/viewgame-1445-secretary.html"),
      "utf8",
    );

    const subject = new GamesApiImpl();
    const result = subject.parseGameDetails(1445, response);
    expect(result).toMatchSnapshot();
  });
  it("should parse a viewgame response 2", () => {
    const response = fs.readFileSync(
      path.join(import.meta.dirname, "test/viewgame-3227-slat.html"),
      "utf8",
    );

    const subject = new GamesApiImpl();
    const result = subject.parseGameDetails(3227, response);
    expect(result).toMatchSnapshot();
  });
});
