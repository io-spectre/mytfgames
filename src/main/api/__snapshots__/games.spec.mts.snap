// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Games API > should parse a list response 1`] = `
[
  {
    "id": 314,
    "lastUpdate": "2013-11-11T00:00:00Z",
    "name": "The uninvited guest",
  },
  {
    "id": 317,
    "lastUpdate": "2013-03-13T00:00:00Z",
    "name": "Cromexx Funhouse for NWN 1",
  },
  {
    "id": 66,
    "lastUpdate": "2013-01-12T00:00:00Z",
    "name": "Vampire Fate",
  },
]
`;

exports[`Games API > should parse a viewgame response 1`] = `
{
  "authors": [
    {
      "id": 38045,
      "name": "Deedee",
    },
    {
      "id": 51960,
      "name": "psy",
    },
  ],
  "id": 1445,
  "lastUpdate": "2024-05-12T00:00:00Z",
  "name": "Secretary",
  "releaseDate": "2017-10-01T18:57:00Z",
  "transformationThemes": [
    {
      "abbreviation": "BE",
      "id": 4,
      "name": "Breast Enlargement",
    },
    {
      "abbreviation": "Grow",
      "id": 7,
      "name": "Growth/Expansion",
    },
    {
      "abbreviation": "Invol",
      "id": 8,
      "name": "Involuntary",
    },
    {
      "abbreviation": "M2F",
      "id": 9,
      "name": "Male to Female (Genital Changes)",
    },
    {
      "abbreviation": "Fem",
      "id": 45,
      "name": "Feminization",
    },
    {
      "abbreviation": "Slow",
      "id": 40,
      "name": "Slow Transformation",
    },
    {
      "abbreviation": "MC",
      "id": 10,
      "name": "Mental Changes",
    },
    {
      "abbreviation": "Bimbo",
      "id": 15,
      "name": "Bimbo",
    },
    {
      "abbreviation": "Herm",
      "id": 16,
      "name": "Hermaphrodite",
    },
    {
      "abbreviation": "Corr",
      "id": 39,
      "name": "Corruption",
    },
    {
      "abbreviation": "SheM",
      "id": 19,
      "name": "Shemale",
    },
    {
      "abbreviation": "Vol",
      "id": 23,
      "name": "Voluntary",
    },
    {
      "abbreviation": "Sissy",
      "id": 44,
      "name": "Sissification",
    },
  ],
  "versions": [
    {
      "downloads": [
        {
          "link": "https://www.mediafire.com/file/hl1azwap3l1xy5x/secretary1013.rar/file",
          "name": "mediafire.com",
          "note": "MediaFire",
        },
        {
          "link": "https://mega.nz/file/sOJzXLYT#V00E8-ZClD9Ej9G7Oloavt41F8NJQ_0GQm2ppvdXUxo",
          "name": "mega.nz",
          "note": "Mega",
        },
      ],
      "version": "1.0.1.3",
    },
    {
      "downloads": [
        {
          "link": "https://www.mediafire.com/file/rwlg3u68598r1hb/secretary1005.rar/file",
          "name": "mediafire.com",
          "note": "MediaFire",
        },
        {
          "link": "https://mega.nz/file/xbQWjQRD#yxVeWBth7AzAeX03h6GJ6_ptd7j8mEsAdQZb1Mf_UW4",
          "name": "mega.nz",
          "note": "Mega",
        },
      ],
      "version": "1.0.0.5",
    },
  ],
}
`;

exports[`Games API > should parse a viewgame response 2 1`] = `
{
  "authors": [
    {
      "id": 160407,
      "name": "MariaMod",
    },
  ],
  "id": 3227,
  "lastUpdate": "2024-06-14T00:00:00Z",
  "name": "S.L.A.T. Foundation",
  "releaseDate": "2024-03-17T00:00:00Z",
  "transformationThemes": [
    {
      "abbreviation": "BE",
      "id": 4,
      "name": "Breast Enlargement",
    },
    {
      "abbreviation": "M2F",
      "id": 9,
      "name": "Male to Female (Genital Changes)",
    },
    {
      "abbreviation": "Fem",
      "id": 45,
      "name": "Feminization",
    },
    {
      "abbreviation": "MC",
      "id": 10,
      "name": "Mental Changes",
    },
    {
      "abbreviation": "Poss",
      "id": 12,
      "name": "Possession",
    },
    {
      "abbreviation": "Shrink",
      "id": 14,
      "name": "Shrinking",
    },
    {
      "abbreviation": "Corr",
      "id": 39,
      "name": "Corruption",
    },
    {
      "abbreviation": "SheM",
      "id": 19,
      "name": "Shemale",
    },
    {
      "abbreviation": "Vol",
      "id": 23,
      "name": "Voluntary",
    },
    {
      "abbreviation": "Sissy",
      "id": 44,
      "name": "Sissification",
    },
  ],
  "versions": [
    {
      "downloads": [
        {
          "link": "https://MilkyNail.site/slat.zip",
          "name": "milkynail.site",
          "note": "Online/Offline version with pics",
        },
      ],
      "version": "All versions",
    },
  ],
}
`;
